### PR TITLE
VCL_REGEX

### PR DESCRIPTION
**Pull request by Dridi**
I wrote this patch myself, ended up not reviewing a previous submission.

There are open questions in VRT where part of the API uses `VCL_REGEX` and others use `void *`.

### CODE REVIEW

**Comment by Dridi** [comment]
This pull request introduces a REGEX type for VCL.

Today regular expressions can only be literals, making it impossible to use built-in functions or operators with a regular expression coming from a VMOD. Conversely a VMOD can also take a REGEX from a STRING literal without having to deal with compilation since that would be handled automatically, putting VMOD functions on par with `regsub[all]`.

The first patch introduces a REGEX type and allows regular expressions to be expressions in addition to literals.

The second patch teaches VMODs to interact with the REGEX type and adds some coverage via vmod_debug.

The pull request is still a draft because today we use `void *` to pass VREs around and we might want to unvoid the VRT API.

**Comment by Dridi** [comment]
Bugwash consensus: go all the way and include documentation (I will undraft the pull request then).

Also requested:

- rename `debug.regex()` to `debug.just_return_regex()`
- comment about parentheses in the test case

**Comment by Dridi** [comment]
Rebased against master, with review items taken care off. I went the extra mile and adopted the REGEX type in vmod_cookie, bringing an overall negative diff stat to the whole pull request.

Ready for review.

**Comment by Dridi** [comment]
Force-pushed, I had the wrong priv in d990d0ac5fc6ec492abdf878c8b1aa802f6d26da's commit log. No change in the diff itself.

**Comment by Dridi** [comment]
Ready for review, with an update to the docs this time.

File: include/vrt.h
Comment (nigoroll) [inline include/vrt.h:459]: :-1:  on removing `VRT_re_*` from VRT.
This will break vmods building VREs dynamically.
According to my vmod library, this issue is real.

File: bin/varnishtest/tests/c00103.vtc
Comment (nigoroll) [inline bin/varnishtest/tests/c00103.vtc:11]: what happens when a dynamic string is used as a REGEX argument?

**Review by nigoroll — CHANGES_REQUESTED** [review]
see comments.
Overall I am still very much in favor, but there is one show stopper.

File: bin/varnishtest/tests/c00103.vtc
Comment (Dridi) [reply bin/varnishtest/tests/c00103.vtc:11]: With this, only CSTR tokens can be used, so what happens is a compile-time error.

File: include/vrt.h
Comment (Dridi) [reply include/vrt.h:459]: In that case you are asking for a panic if the dynamically-built regex has a syntax error.

FWIW you have `vre.h` to work with such regular expressions. That's what vmod_cookie was using and it was not affected by the privatization of that part of the API. I would rather reconsider how those VMODs are operating today.

Unless we don't want to offer VCL scope and compile-time checks to the REGEX type.

Comment (nigoroll) [reply include/vrt.h:459]: the panic can be avoiding by another `VRE_compile()` before the `VRT_re_init()`, but I had a closer look and now agree that the use cases in vmods are [either wrong](https://github.com/varnish/varnish-modules/blob/f26e057727c6e4ff3c042dae19cfd64e75a95236/src/vmod_header.c#L254) or [motivated by convenience](https://github.com/Dridi/libvmod-querystring/blob/06f8206c91842d44f881c1a343c264312d3a36a7/src/vmod_querystring.c#L598) (examples chosen randomly ;) )

So I agree with you now.

File: bin/varnishtest/tests/c00103.vtc
Comment (nigoroll) [reply bin/varnishtest/tests/c00103.vtc:11]: I was trying to imply that we should have a test ;)

Comment (nigoroll) [reply bin/varnishtest/tests/c00103.vtc:11]: thank you, our comments have crossed.

**Comment by bsdphk** [comment]
OK with me.

**Comment by Dridi** [comment]
Rebased against master to solve a conflict in `vrt.h`. Waiting for CI to complete before merging.

**Pull request merged**
